lib/sles4sap/sap_deployment_automation_framework/basetest.pm

```diff
@@ -95,6 +96,34 @@ sub full_cleanup {
 
 sub post_fail_hook {
     record_info('Post fail', 'Executing post fail hook');
+    if (testapi::is_serial_terminal()) {
+        if ($serial_regexp_playbook) {
+            # In case playbook times out, it will keep occupying the command line,
+            # therefore we need to press Ctrl+c to terminate the process.
+            # Note: please do not use '> ' (or $testapi::distri->{serial_term_prompt})
+            # as there are lots of '> ' in the output of playbook
+            # 'qr/-\d+-/' is usually the last output from playbook execution
+            unless (wait_serial(qr/-\d+-/)) {
+                type_string('', terminate_with => 'ETX');
+                # Wait for process returns
+                wait_serial(qr/-\d+-/);
+                type_string("\n");
+                record_info('Terminated playbook process');
+            }
+            $serial_regexp_playbook = 0;
+        }
+        else {
+            # Teminate the process when other scripts timed out
+            unless (wait_serial($testapi::distri->{serial_term_prompt})) {
+                type_string('', terminate_with => 'ETX');
+                # Wait for process returns
+                wait_serial(qr/-\d+-/);
+                type_string("\n");
```

Comment:
It looks good, but I don't like the duplicated code in L107-L110 and then on L118-L120. Perhaps we can refactor that somehow. It could go into a helper/not-exported function within the module, or something along these lines:
```
if (testapi::is_serial_terminal()) {
    my $match_re = $testapi::distri->{serial_term_prompt};
    $match_re = qr/-\d+-/ if ($serial_regexp_playbook);
    unless (wait_serial($match_re)) {
        type_string('', terminate_with => 'ETX');
        # Wait for process returns
        wait_serial(qr/-\d+-/);
        type_string("\n");
        if ($serial_regexp_playbook) { record_info('Terminated playbook process') }
        else { record_info('Terminated other script process') }
}
```
WDYT?

Reply:
Updated and the new VRs are good please help to review again:
https://openqaworker15.qa.suse.cz/tests/302930#step/deploy_hanasr/461 (failed on playbook timed out) http://openqaworker15.qa.suse.cz/tests/302928#step/deploy_workload_zone/458 (failed on other script `assert_script_run("sleep 200")` this line was added on purpose)
  
Also added upload logs when playbook timed out: see https://openqaworker15.qa.suse.cz/tests/302930/logfile?filename=deploy_hanasr-playbook_04_00_00_db_install.txt